spring-kafka/src/main/java/org/springframework/kafka/support/DefaultKafkaHeaderMapper.java

```diff
@@ -0,0 +1,232 @@
+/*
+ * Copyright 2017 the original author or authors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.springframework.kafka.support;
+
+import java.io.IOException;
+import java.text.MessageFormat;
+import java.util.ArrayList;
+import java.util.Arrays;
+import java.util.HashMap;
+import java.util.Iterator;
+import java.util.List;
+import java.util.Map;
+
+import org.apache.commons.logging.Log;
+import org.apache.commons.logging.LogFactory;
+import org.apache.kafka.common.header.Header;
+import org.apache.kafka.common.header.Headers;
+import org.apache.kafka.common.header.internals.RecordHeader;
+
+import org.springframework.messaging.MessageHeaders;
+import org.springframework.messaging.support.HeaderMapper;
+import org.springframework.util.Assert;
+import org.springframework.util.ClassUtils;
+import org.springframework.util.PatternMatchUtils;
+
+import com.fasterxml.jackson.core.JsonProcessingException;
+import com.fasterxml.jackson.databind.ObjectMapper;
+
+/**
+ * Header mapper for kafka.
+ *
+ * @author Gary Russell
+ * @since 2.0
+ *
+ */
+public class DefaultKafkaHeaderMapper implements HeaderMapper<Headers> {
+
+	private static final Log logger = LogFactory.getLog(DefaultKafkaHeaderMapper.class);
+
+	private static final String JSON_TYPES = "spring_json_header_types";
+
+	private final ObjectMapper objectMapper;
+
+	private final List<SimplePatternBasedHeaderMatcher> matchers = new ArrayList<>();
+
+	/**
+	 * Construct an instance with the default object mapper and header patterns for
+	 * outbound headers; all inbound headers are mapped.
+	 * @see #DefaultKafkaHeaderMapper(ObjectMapper)
+	 */
+	public DefaultKafkaHeaderMapper() {
+		this(new ObjectMapper());
+	}
+
+	/**
+	 * Construct an instance with the provided object mapper and default header patterns
+	 * for outbound headers; all inbound headers are mapped. The patterns are applied in
+	 * order, stopping on the first match (positive or negative). Patterns are negated by
+	 * preceding them with "!". The default pattern list is
+	 * {@code "!id", "!timestamp" and "*"}.
+	 * @param objectMapper the object mapper.
+	 * @see org.springframework.util.PatternMatchUtils#simpleMatch(String, String)
+	 */
+	public DefaultKafkaHeaderMapper(ObjectMapper objectMapper) {
```

Comment:
Several reasons:

1. Simple
1. Easy to detect it was originated by a Spring app (presence of JSON_TYPES header)
1. Easy to document what a non-Spring app would need to do.
1. If we supported multiple types, we'd need some way to identify the type to create on the consumer side (and do all that work ourselves rather than letting Jackson do it).
1. There is no standard way to convey that type information.
1. The user can use his own scheme using messaging Message<?> headers with `byte[]` values, which are sent unchanged, or by writing his own mapper.